src/translate.py

```diff
@@ -3841,23 +4140,100 @@ def process_instr(instr: Instruction) -> None:
+            set_reg(Register("cr0_eq"), arch.instrs_ppc_compare[mnemonic](args, "=="))
+            set_reg(Register("cr0_gt"), arch.instrs_ppc_compare[mnemonic](args, ">"))
+            set_reg(Register("cr0_lt"), arch.instrs_ppc_compare[mnemonic](args, "<"))
+            set_reg(Register("cr0_so"), Literal(0))
+
         elif mnemonic in arch.instrs_no_dest:
             stmt = arch.instrs_no_dest[mnemonic](args)
             to_write.append(stmt)
 
-        elif mnemonic in arch.instrs_destination_first:
+        elif mnemonic.rstrip(".") in arch.instrs_destination_first:
             target = args.reg_ref(0)
-            val = arch.instrs_destination_first[mnemonic](args)
+            val = arch.instrs_destination_first[mnemonic.rstrip(".")](args)
             # TODO: IDO tends to keep variables within single registers. Thus,
             # if source = target, maybe we could overwrite that variable instead
             # of creating a new one?
             set_reg(target, val)
             mn_parts = mnemonic.split(".")
-            if (len(mn_parts) >= 2 and mn_parts[1] == "d") or mnemonic == "ldc1":
+            if mnemonic.endswith("."):
+                # PPC instructions suffixed with . set condition bits (CR0) based on the result value
+                target_reg = args.reg(0)
+                set_reg(
+                    Register("cr0_eq"),
+                    BinaryOp.icmp(target_reg, "==", Literal(0, type=target_reg.type)),
+                )
+                # Use manual casts for cr0_gt/cr0_lt so that the type of target_reg is not modified
+                # until the resulting bit is .use()'d.
+                set_reg(
+                    Register("cr0_gt"),
+                    BinaryOp(
+                        left=Cast(
+                            expr=target_reg,
+                            reinterpret=True,
+                            silent=True,
+                            type=Type.s32(),
+                        ),
+                        op=">",
+                        right=Literal(0),
+                        type=Type.s32(),
+                    ),
+                )
+                set_reg(
+                    Register("cr0_lt"),
+                    BinaryOp(
+                        left=Cast(
+                            expr=target_reg,
+                            reinterpret=True,
+                            silent=True,
+                            type=Type.s32(),
+                        ),
+                        op="<",
+                        right=Literal(0),
+                        type=Type.s32(),
+                    ),
+                )
+                # TODO: Implement overflow checking
+                set_reg(Register("cr0_so"), ErrorExpr(f"overflow of {target}"))
+
+            elif (len(mn_parts) >= 2 and mn_parts[1] == "d") or mnemonic == "ldc1":
                 set_reg(target.other_f64_reg(), SecondF64Half())
 
+        elif mnemonic in arch.instrs_load_update:
```

Comment:
With the load/store_update instructions, it could potentially be nice if we were somehow able to share EvalOnce for the addition created by the handler in arch.instrs_load_update[], and for the update. Particularly for x versions, since the imm ones likely won't end up needing a temp. Just theorizing though, I don't know if this would help in practice. I also guess ux instructions are probably quite rare in practice. (Edit: just checked and there's a grand total of 0 of them in the instruction tracking issue. :) So let's maybe not worry too much here.)

We could potentially get rid of this whole category of instructions by asm-pattern-replacing load-update's with update + load or load + update (with scratch registers if needed). (There's a the risk that addi + load with offset 0 won't give as good results as a single load with correct offset; I don't know how much pattern detection we do in that regard.)

Let's assert that rA != rB. ("If rA = 0 or rA = rD, the instruction form is invalid.")

Reply:
I added the assertion, but didn't make any of the suggested improvements.